library/alloc/src/raw_vec.rs

```diff
@@ -77,6 +82,237 @@ pub(crate) struct RawVec<T, A: Allocator = Global> {
+        // None.
+        if cap == 0 {
+            unsafe { self.alloc.deallocate(ptr, layout) };
+            self.ptr = unsafe { Unique::new_unchecked(ptr::without_provenance_mut(align)) };
+            self.cap = Cap::ZERO;
+        } else {
+            let ptr = unsafe {
+                // `Layout::array` cannot overflow here because it would have
+                // overflowed earlier when capacity was larger.
+                let new_size = size.unchecked_mul(cap);
+                let new_layout = Layout::from_size_align_unchecked(new_size, layout.align());
+                self.alloc
+                    .shrink(ptr, layout, new_layout)
+                    .map_err(|_| AllocError { layout: new_layout, non_exhaustive: () })?
+            };
+            // SAFETY: if the allocation is valid, then the capacity is too
+            unsafe {
+                self.set_ptr_and_cap(ptr, cap);
+            }
+        }
+        Ok(())
+    }
+
+    #[inline]
+    pub const fn new_in(alloc: A, align: usize) -> Self {
+        let ptr = unsafe { Unique::new_unchecked(ptr::without_provenance_mut(align)) };
+        // `cap: 0` means "unallocated". zero-sized types are ignored.
+        Self { ptr, cap: Cap::ZERO, alloc }
+    }
+
+    #[inline]
+    fn try_allocate_in(
+        capacity: usize,
+        init: AllocInit,
+        alloc: A,
+        size: usize,
+        align: usize,
+    ) -> Result<Self, TryReserveError> {
+        // Don't allocate here because `Drop` will not deallocate when `capacity` is 0.
+        if size == 0 || capacity == 0 {
+            Ok(Self::new_in(alloc, align))
+        } else {
+            // We avoid `unwrap_or_else` here because it bloats the amount of
+            // LLVM IR generated.
+            let layout = match layout_array(capacity, size, align) {
+                Ok(layout) => layout,
+                Err(_) => return Err(CapacityOverflow.into()),
+            };
+
+            if let Err(err) = alloc_guard(layout.size()) {
+                return Err(err);
+            }
+
+            let result = match init {
+                AllocInit::Uninitialized => alloc.allocate(layout),
+                #[cfg(not(no_global_oom_handling))]
+                AllocInit::Zeroed => alloc.allocate_zeroed(layout),
+            };
+            let ptr = match result {
+                Ok(ptr) => ptr,
+                Err(_) => return Err(AllocError { layout, non_exhaustive: () }.into()),
+            };
+
+            // Allocators currently return a `NonNull<[u8]>` whose length
+            // matches the size requested. If that ever changes, the capacity
+            // here should change to `ptr.len() / mem::size_of::<T>()`.
+            Ok(Self { ptr: Unique::from(ptr.cast()), cap: unsafe { Cap(capacity) }, alloc })
+        }
+    }
+
+    #[inline]
+    #[rustc_no_mir_inline]
+    fn drop_if_needed(&mut self, size: usize, align: usize) {
```

Comment:
That would be more MIR on the polymorphic side.

Reply:
If the current state of this PR looks good, I'll try other ideas to clean this up; things like a `const` for the layout of each type. I'm not sure if that nets out to a benefit, because that const will be instantiated a zillion times. But it should be propagated into the call operands, I think? That might be good?